src/storage/cache/rpc/workerpool.rs

```diff
@@ -0,0 +1,242 @@
+    worker_queue: Vec<Worker>,
+}
+
+impl Debug for WorkerPool {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("WorkerPool")
+            .field("max_workers", &self.max_workers)
+            .field("max_waiting_jobs", &self.max_waiting_jobs)
+            .field("worker_queue_len", &self.worker_queue.len())
+            .finish_non_exhaustive()
+    }
+}
+
+impl WorkerPool {
+    /// Create a new worker pool, which contains a number of max workers and max waiting jobs.
+    #[must_use]
+    pub fn new(max_workers: usize, max_waiting_jobs: usize) -> Self {
+        let (job_sender, job_receiver) = flume::bounded::<JobImpl>(max_waiting_jobs);
+        let mut worker_queue = Vec::new();
+
+        // In current implementation, we create a fixed number of workers.
+        let receiver = Arc::new(job_receiver);
+        for _ in 0..max_workers {
+            let worker = Worker::new(Arc::clone(&receiver));
+            worker_queue.push(worker);
+        }
+
+        // let receiver_clone = receiver.clone();
+        // tokio::task::spawn(async move {
+        //     Self::dispatch_workers(max_workers, receiver_clone).await;
+        // });
+
+        Self {
+            max_workers,
+            max_waiting_jobs,
+            job_sender: Arc::new(job_sender),
+            worker_queue,
+        }
+    }
+
+    /// Submit a job to the worker pool asynchronously, and return a future that resolves when the job is completed.
+    pub fn submit_job(&self, job: JobImpl) -> Result<(), RpcError<String>> {
+        // If current running worker count is more than max workers, and the job queue is full, return false.
+        if self.job_sender.is_full() {
+            return Err(RpcError::InternalError("Job queue is full".to_owned()));
+        }
+
+        // Submit the job to the job queue.
+        self.job_sender
+            .try_send(job)
+            .map_err(|_foo| RpcError::InternalError("Failed to submit job".to_owned()))?;
+
+        Ok(())
+    }
+
+    /// Submit a job to the worker pool synchronously, and block until the job is completed.
+    /// Other process will be blocked until the job is completed.
+    /// If all job try to submit the job, the process will be blocked.
+    pub fn submit_job_wait(&self, job: JobImpl) -> Result<(), RpcError<String>> {
+        // Submit the job to the job queue.
+        self.job_sender
+            .send(job)
+            .map_err(|_foo| RpcError::InternalError("Failed to submit job".to_owned()))?;
+
+        Ok(())
+    }
+
+    /// Wait for all workers to complete.
+    pub async fn wait_for_completion(&self) {
+        // Check the sender channel is empty and all workers are completed.
+        while !self.job_sender.is_empty() {
+            tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
+        }
+    }
+
+    /// Delete the worker pool and shutdown all workers.
+    pub fn shutdown(&self) {
+        debug!("Shutting down worker pool...");
```

Comment:
Not finished yet?

Reply:
This is the legacy message for debug, all workers and channels are automatically dropped by `Drop`.